File: nupkgs/OpenSync.Sdk/SyncStream.cs
using OpenSync.Sdk.Events;

namespace OpenSync.Sdk;

public class SyncStream
{
    private readonly OpenSyncClient _client;
    private readonly string _uniqueName;

    public event EventHandler<MessageEventArgs>? OnMessage;

    public SyncStream(OpenSyncClient client, string uniqueName)
    {
        _client = client;
        _uniqueName = uniqueName;
        _client.OnEvent += HandleEvent;
    }

    private void HandleEvent(object? sender, SyncEventArgs args)
    {
        if (args.Event == "stream_message" && args.ObjectId == _uniqueName)
            OnMessage?.Invoke(this, new MessageEventArgs(args.Data));
    }

    public async Task PublishAsync(object data)
    {
        var json = System.Text.Json.JsonSerializer.Serialize(data);
        await _client.HttpPostAsync<object>($"/streams/{_uniqueName}/messages", new { data = json });
    }

    public async Task SubscribeAsync() => await _client.SubscribeAsync("stream", _uniqueName);
}

File: clients/OpenSync.Client.Js/src/SyncStream.ts
import { OpenSyncClient, SyncEvent } from './OpenSyncClient';

export class SyncStream {
  private client: OpenSyncClient;
  private name: string;

  constructor(client: OpenSyncClient, name: string) {
    this.client = client;
    this.name = name;
  }

  async subscribe(): Promise<void> {
    await this.client.subscribe('stream', this.name);
  }

  async publish(data: any): Promise<void> {
    await this.client.httpPost(`/streams/${this.name}/messages`, { data });
  }

  onMessage(handler: (data: any) => void): void {
    this.client.on('stream_message', (event: SyncEvent) => {
      if (event.object_id === this.name) handler(event.data);
    });
  }
}
